File: cliente.py
import threading
import socket

def main():
    cliente = socket.socket(socket.AF_INET, socket.SOCK_STREAM)

    try:
        cliente.connect(("localhost", 1212))
    except:
        print("Não foi possivel se conectar ao servidor! \n")
        return 

    usuario = input("Usuário: ")
    print(f'\n{usuario} Conectado. \nQuando desejar encerrar a conexão, digite "exit". \n')

    thread1 = threading.Thread(target=receber_mensagens, args=[cliente])
    thread2 = threading.Thread(target=enviar_mensagens, args=[cliente, usuario])

    thread1.start()
    thread2.start()
    
    thread1.join()
    thread2.join()

def receber_mensagens(cliente):
    while True:
        try:
            mensagem = cliente.recv(2048).decode('utf-8')
            if mensagem:
                print(f"\n{mensagem}")
            else:
                print("\nConexão encerrada pelo servidor.")
                break
        except:
            break

def enviar_mensagens(cliente,usuario):
    while True:
        try:
            mensagem = input("\n ")
            cliente.send(f'<{usuario}> {mensagem}\n'.encode("utf-8"))
            if mensagem.lower() == "exit":
                print("Encerrando a conexão...")
                cliente.close()
                break
        except Exception as e:
            print(f"\nErro inesperado: {e}")
            break

if __name__ == "__main__":
    main()

File: Cliente.java


import java.io.IOException;
import java.io.PrintWriter;
import java.net.Socket;
import java.net.UnknownHostException;
import java.util.Scanner;

class Receber_mensagens extends Thread {
    private Socket cliSocket;

    public Receber_mensagens(Socket cliSocket) {
        this.cliSocket = cliSocket;
    }

    @Override
    public void run() {
        while (true) {
            try {
                Scanner scanner = new Scanner(cliSocket.getInputStream());
                while (scanner.hasNextLine()) {
                    System.out.println("Servidor: " + scanner.nextLine());
                    System.out.print("\n");
                }
                scanner.close();
            } catch (Exception e) {
                System.out.print("\nErro inesperado: " + e);
                break;
            }
        }
    }   
}

class Enviar_mensagens extends Thread {
    private Socket cliSocket;
    private String usuario;

    public Enviar_mensagens(Socket cliSocket, String usuario) {
        this.cliSocket = cliSocket;
        this.usuario = usuario;
    }

    @Override
    public void run() {
        try {
            System.out.println("Digite mensagens para o servidor (ou 'exit' para encerrar): \n");

            Scanner teclado = new Scanner(System.in);
            PrintWriter saida = new PrintWriter(cliSocket.getOutputStream(), true);
            while (teclado.hasNextLine()) {
                String mensagem = teclado.nextLine();
                System.out.print("\n");
                if (mensagem.equalsIgnoreCase("exit")) {
                    System.out.println("\nEncerrando conexão...");
                    teclado.close();
                    cliSocket.close();
                    break;
                }
                saida.printf("<%s> %s", usuario, mensagem);
            }
        } catch (Exception e) {
            System.out.print("\nErro inesperado: " + e);
        }
    }   
}

public class Cliente {
    public static void main(String[] args) throws UnknownHostException, IOException, InterruptedException {
        try {
            Socket cliente = new Socket("localhost", 1212);
            Scanner teclado = new Scanner(System.in);
            System.out.println("Conectado ao servidor!");
            System.out.print("Usuário: ");
            String usuario = teclado.nextLine();
            System.out.println(usuario + " conectado.");

            Thread recebeThread = new Receber_mensagens(cliente);
            recebeThread.start();

            Thread enviaThread = new Enviar_mensagens(cliente, usuario);
            enviaThread.start();

            recebeThread.join();
            enviaThread.join();
            
            teclado.close();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

